src/cucumber.rs

```diff
@@ -153,7 +154,21 @@ impl<W: World> Cucumber<W> {
         let mut stream = runner.run();
 
         while let Some(event) = stream.next().await {
+            let ret = {
+                if let crate::event::CucumberEvent::Finished(ref result) = event {
+                    Some(result.clone())
+                } else {
+                    None
+                }
+            };
+
             self.event_handler.handle_event(event);
+
+            if let Some(res) = ret {
+                return res
+            }
         }
+
+        unimplemented!("Programmer Error - run _must end_ with the Finished Event")
```

Comment:
another quirk of the implementation: theoretically the stream could end without the proper result being yielded and then we'd run into this here. not sure how to make this nicer...

Reply:
It is a bit of a code smell, and at a glance I don't see a way to make the logic nicer. Maybe the ref change will make it more flexible.

In any case, if you need to keep it, I'd change this to `panic!("Invariant broken: no Finish event found before stream ended")` or something similar.